shutter/low-level.md

```diff
@@ -0,0 +1,307 @@
+| ORDER    | The order of groups G1 and G2 | 21888242871839275222246405745257275088548364400416034343698204186575808495617 |
+
+The following functions are considered prerequisites:
+
+| Function                       | Description                                                                                     |
+| ------------------------------ | ----------------------------------------------------------------------------------------------- |
+| keccak256(bytes) -> Block      | The keccak-256 hash function                                                                    |
+| pairing(G1, G2) -> GT          | The pairing function specified in [EIP-197](https://eips.ethereum.org/EIPS/eip-197)             |
+| g1_scalar_base_mult(int) -> G1 | Multiply the generator of G1 by a scalar                                                        |
+| g2_scalar_base_mult(int) -> G2 | Multiply the generator of G2 by a scalar                                                        |
+| gt_scalar_mult(GT, int)        | Multiply an element of GT by a scalar                                                           |
+| encode_g1(G1) -> bytes         | Encode an element of G1 according to [EIP-197](https://eips.ethereum.org/EIPS/eip-197#encoding) |
+| encode_g2(G2) -> bytes         | Encode an element of G2 according to [EIP-197](https://eips.ethereum.org/EIPS/eip-197#encoding) |
+
+### Helper Functions
+
+```Python
+def hash_block_to_int(block: Block) -> int:
+    h = keccak256(block)
+    i = int.from_bytes(h, "big")
+    return i % ORDER
+
+def hash_gt_to_block(preimage: GT) -> Block:
+    b = encode_gt(preimage)
+    return hash_bytes_to_block(b)
+
+def hash_bytes_to_block(preimage: bytes) -> Block:
+    return keccak256(preimage)
+
+def encode_gt(value: GT) -> bytes:
+    pass  # TODO
+
+def xor_blocks(block1: Block, block2: Block) -> Block:
+    return Block(bytes(b1 ^ b2 for b1, b2 in zip(block1, block2)))
+
+def pad_and_split(b: bytes) -> Sequence[Block]:
+    # pad according to PKCS #7
+    n = 32 - len(b) % 32
+    padded = b + n * bytes([n])
+    return [padded[i:i + 32] for i in range(0, len(padded), 32)]
+
+def unpad_and_join(blocks: Sequence[Block]) -> bytes:
+    # unpad according to PKCS #7
+    if len(blocks) == 0:
+        return b""
+    last_block = blocks[-1]
+    n = last_block[-1]
+    assert 0 < n <= 32, "invalid padding length"
+    return b"".join(blocks)[:-n]
+
+def compute_block_keys(sigma: Block, n: int) -> Sequence[Block]:
+    suffix_length = max(n.bit_length() + 7) // 8, 1)
+    suffixes = [n.to_bytes(suffix_length, "big")]
+    preimages = [sigma + suffix for suffix in suffixes]
+    keys = [hash_bytes_to_block(preimage) for preimage in preimages]
+    return keys
+```
+
+### Encryption and Decryption
+
+```Python
+def encrypt(message: bytes, identity: G1, eon_key: G2, sigma: Block) -> EncryptedMessage:
```

Comment:
Where are this functions used? Could you link this section with expected usage on the contracts above? For example specify that the argument `encryptedTransaction` from `submitEncryptedTransaction(eon, identityPrefix, encryptedTransaction, gasLimit)` is the output of this function.